cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/CentralOAuthServiceImpl.java

```diff
@@ -288,29 +293,23 @@ public Boolean revokeToken(final Token token) {
 
     @Override
     public Boolean revokeClientTokens(final String clientId, final String clientSecret) {
-        final OAuthRegisteredService service = getRegisteredService(clientId);
-        if (service == null) {
-            LOGGER.error("OAuth Registered Service could not be found for clientId : {}", clientId);
-            return Boolean.FALSE;
-        }
-        if (!service.getClientSecret().equals(clientSecret)) {
-            LOGGER.error("Invalid client secret");
-            return Boolean.FALSE;
-        }
 
         final Collection<RefreshToken> refreshTokens = tokenRegistry.getClientTokens(clientId, RefreshToken.class);
-        for (final RefreshToken token : refreshTokens) {
-            LOGGER.debug("Revoking refresh token : {}", token.getId());
-            ticketRegistry.deleteTicket(token.getTicket().getId());
-        }
-
         final Collection<AccessToken> accessTokens = tokenRegistry.getClientTokens(clientId, AccessToken.class);
-        for (final AccessToken token : accessTokens) {
-            LOGGER.debug("Revoking access token : {}", token.getId());
-            ticketRegistry.deleteTicket(token.getTicket().getId());
-        }
 
-        return Boolean.TRUE;
+        Token token;
+        if (!refreshTokens.isEmpty()) {
+            token = refreshTokens.iterator().next();
+            LOGGER.debug("Revoking refresh token : {}", token.getId());
+        } else if (!accessTokens.isEmpty()) {
+            token = accessTokens.iterator().next();
+            LOGGER.info("Revoking access token : {}", token.getId());
+        } else {
+            // all tokens are cleared
+            return Boolean.TRUE;
+        }
+        ticketRegistry.deleteTicket(token.getTicket().getId());
```

Comment:
### The Fix

The issue is fixed by only revoke one token at a time in the callee:
```java
CentralOAuthServiceImpl.revokeClientTokens
```
and move the for logic (and the service check as a side effect) to the caller:
```java
OAuth20RevokeClientTokensController. handleRequestInternal()
```
`revokeClientTokens()` returns `Boolean.TRUE` if all tokens are cleared and `Boolean.FASLE` if there is more to continue.